### DISCUSSION TITLE
Recommended way to go when using graphql?

### DISCUSSION THREAD

**Discussion by t-zander** [vercel/next.js#51416] [Help]
### Summary

I am really grateful for nextjs 13 and app router. It was really exciting but... then I wanted to use graphql and here is when I got completely lost tbh.

I see multiple options about how to use graphql with nextjs 13.
1. using something lightweight like `graphql-request` and graphql codegen. 
https://github.com/vercel/next.js/issues/49438

The problem is - caching doesn't really work https://github.com/vercel/next.js/issues/49438
although I can use `cache` function it doesn't seem to be fully ready and production ready.

2. Use something like apollo
Revalidate option doesn't work.
Also the package to work with apollo in nexjt 13 is experimental. https://github.com/apollographql/apollo-client-nextjs
https://github.com/apollographql/apollo-client-nextjs/issues/24

I understand that other libs like apollo or graphql request will need time to adapt, but currently I feel completely lost. All the examples are only about using `fetch`, but I think a lot of projects are using graphql and then using it with fetch looks quite ugly tbh.

I would really appreciate if docs could have any kind of direction or examples regarding graphql. For now it's a bit frustrating as some things like cache don't work well and there is no official ways from next.js team.

**Comment by SamKomesarook** [comment]
We haven't tried the packages you linked, however we are using the [standard apollo client](https://github.com/apollographql/apollo-client) in our Next13 project with success. We create our Apollo Clients in util functions, then use those to query or mutate throughout the app. Happy to provide more specific examples, but are there any issues with that one for your use case?

**Comment by knoefel** [comment]
Great question! I'm facing the same situation right now, as we want to move our application to the app folder. @t-zander Have you been able to find a solution in the meantime? 

@leerob Are there any plans for next to provide an official example of how caching can be solved with graphql? Right now this is a blocker for us to switch to the app folder. As i understand the docs simply passing the augmented `fetch` to an graphql adapter, like `graphql-request` won't work:

> Dynamic methods (next/headers, export const POST, or similar) are used and the fetch is a POST request (or uses Authorization or cookie headers)

 Thanks in advance for any feedback.

**Comment by affanshahid** [comment]
I have the same question. Using 'React.cache' provides no way of invalidating the cached requests
